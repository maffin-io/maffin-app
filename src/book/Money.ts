import * as currencies from '@dinero.js/currencies';
import type { Currency as DineroCurrency } from 'dinero.js';
import * as djs from 'dinero.js';

import { toAmountWithScale, moneyToString } from '@/helpers/number';

export default class Money {
  private _raw: djs.Dinero<number>;

  /**
   * A wrapper class for dinero object to reduce verbosity and overhead. The constructor
   * checks for how many decimals the amount has and calculates the scale automatically.
   *
   * @param amount - The amount to be considered in the money object. Can be a float
   * @param currency - The currency
   * @param scale - Number representing custom scale. Default is 0, if passed differently, that one
   * is taken into account
   */
  constructor(amount: number, currency: string, scale?: number) {
    const dineroCurrency = toCurrency(currency);

    if (scale === undefined) {
      const { amount: newAmount, scale: newScale } = toAmountWithScale(amount);
      this._raw = djs.dinero({
        currency: dineroCurrency,
        amount: newAmount,
        scale: newScale,
      });
    } else {
      this._raw = djs.dinero({
        currency: dineroCurrency,
        amount,
        scale,
      });
    }
  }

  get currency(): string {
    return djs.toSnapshot(this._raw).currency.code;
  }

  get raw(): djs.Dinero<number> {
    return this._raw;
  }

  /**
   * Represents the money object as a string with the currency standardizing
   * the scale to 2 as a default. If the scale is passed then it uses that.
   */
  toString(scale = 2): string {
    return djs.toDecimal(
      djs.transformScale(this._raw, scale),
      ({ value, currency }) => `${value} ${currency.code}`,
    );
  }

  /**
   * Represents the money object as a string with the currency standardizing
   * the scale to 2 as a default. If the scale is passed then it uses that.
   *
   * The result is a localized string with the currency as a symbol
   */
  format(scale = 2): string {
    return djs.toDecimal(
      djs.transformScale(this._raw, scale),
      ({ value, currency }) => moneyToString(Number(value), currency.code),
    );
  }

  toNumber(): number {
    return parseFloat(djs.toDecimal(this._raw));
  }

  isNegative(): boolean {
    return djs.isNegative(this._raw);
  }

  isZero(): boolean {
    return djs.isZero(this._raw);
  }

  /**
   * Adds b to the current object and returns a new one with the result
   *
   * @param b - Money to be added
   * @returns - Money with result
   */
  add(b: Money): Money {
    const result = djs.toSnapshot(djs.add(b.raw, this.raw));
    return new Money(result.amount, this.currency, result.scale);
  }

  /**
   * Subtracts b to the current object and returns a new one with the result
   *
   * @param b - Money to be subtracted
   * @returns - Money with result
   */
  subtract(b: Money): Money {
    const result = djs.toSnapshot(djs.subtract(this.raw, b.raw));
    return new Money(result.amount, this.currency, result.scale);
  }

  /**
   * Multiplies b to the current object and returns a new one with the result
   * Initially, we try to multiply without transforming b but sometimes dinerojs
   * returns that it is an invalid amount (I have no idea why). If it fails, then
   * we transform b to `amountWithScale` and multiply with that.
   *
   * @param b - amount to be multiplied to in scale 0
   * @returns - Money with result
   */
  multiply(b: number): Money {
    let result;

    try {
      result = djs.multiply(this.raw, b);
    } catch {
      result = djs.multiply(this.raw, toAmountWithScale(b));
    }

    const snapshot = djs.toSnapshot(result);
    return new Money(snapshot.amount, this.currency, snapshot.scale);
  }

  /**
   * Checks for equality with b by comparing amount and currency
   *
   * @returns - boolean
   */
  equals(b: Money): boolean {
    return djs.equal(this._raw, b._raw);
  }

  /**
   * Converts a Money object to the currency specified
   *
   * @param to - Money object containing the rate to
   */
  convert(
    to: string,
    rate: number,
  ): Money {
    const rates = {
      [to]: {
        ...toAmountWithScale(rate),
      },
    };
    const currency = toCurrency(to);

    try {
      // @ts-ignore
      const result = djs.toSnapshot(djs.convert(this.raw, currency, rates));
      return new Money(result.amount, currency.code, result.scale);
    } catch (e) {
      throw new Error(`Error converting ${this.format()} to ${to} with rates ${JSON.stringify(rates)}: ${e}`);
    }
  }
}

function toCurrency(currency: string): DineroCurrency<number> {
  // @ts-ignore
  let dineroCurrency = currencies[currency];
  if (!dineroCurrency) {
    dineroCurrency = {
      code: currency,
      base: 10,
      exponent: 2,
    };
  }
  return dineroCurrency;
}
