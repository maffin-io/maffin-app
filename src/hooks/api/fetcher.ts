export default function fetcher(f: () => Promise<any>, key: unknown) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${JSON.stringify(key)}: ${end - start}ms`);
    return r;
  };
}
