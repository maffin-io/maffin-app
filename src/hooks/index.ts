export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
