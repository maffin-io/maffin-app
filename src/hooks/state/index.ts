export { useTheme } from './useTheme';
export { useOnline } from './useOnline';
