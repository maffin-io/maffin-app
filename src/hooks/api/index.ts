import { DateTime } from 'luxon';
import { mutate, SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity } from '@/book/entities';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import type { PriceDBMap } from '@/book/prices';
import type { AccountsMap } from '@/types/book';
import type { Account, Split, Transaction } from '@/book/entities';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodity(guid: string): SWRResponse<Commodity> {
  const key = `/api/commodities/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.findOneByOrFail({ guid }), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.find(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (commodity: Commodity) => mutate(
        `/api/commodities/${commodity.guid}`,
        commodity,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccount(guid: string): SWRResponse<Account> {
  const key = `/api/accounts/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    Object.values(result.data as AccountsMap).forEach(
      (account: Account) => mutate(`/api/accounts/${account.guid}`, account, { revalidate: false }),
    );
  }

  return result;
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: prices } = useSWRImmutable(
    '/api/prices',
    fetcher(() => queries.getPrices({}), '/api/prices'),
  );

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    (accounts && prices) ? key : null,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, prices),
      key,
    ),
  );

  return result;
}

export function useInvestment(guid: string): SWRResponse<InvestmentAccount> {
  const key = `/api/investments/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestment(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useInvestments(): SWRResponse<InvestmentAccount[]> {
  const key = '/api/investments';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (investment: InvestmentAccount) => mutate(
        `/api/investments/${investment.account.guid}`,
        investment,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(c: Commodity | undefined): SWRResponse<PriceDBMap> {
  const key = `/api/prices/${c?.guid}`;
  return useSWRImmutable(
    c?.guid ? key : null,
    fetcher(async () => queries.getPrices({ from: c }), key),
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
