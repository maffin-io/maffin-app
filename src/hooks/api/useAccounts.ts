import {
  useQuery,
  UseQueryResult,
} from '@tanstack/react-query';

import { Account } from '@/book/entities';
import fetcher from './fetcher';

export function useAccount(guid: string): UseQueryResult<Account | undefined> {
  const result = useAccounts<Account | undefined>(
    (data => data.find(a => a.guid === guid)),
  );

  return result;
}

export function useAccounts<TData = Account[]>(
  select?: (data: Account[]) => TData,
): UseQueryResult<TData> {
  const result = useQuery({
    queryKey: [...Account.CACHE_KEY],
    queryFn: fetcher(
      async () => {
        const accounts = await Account.find();
        const root = accounts.find(a => a.type === 'ROOT') as Account;
        setAccountPaths(root, accounts);
        return accounts;
      },
      Account.CACHE_KEY,
    ),
    select,
    networkMode: 'always',
  });

  return result;
}

function setAccountPaths(current: Account, accounts: Account[]) {
  const parent = accounts.find(a => a.guid === current.parentId);
  if (!parent || parent.type === 'ROOT') {
    current.path = current.name;
  } else {
    current.path = `${parent.path}:${current.name}`;
  }

  current.childrenIds.forEach(childId => {
    const account = accounts.find(a => a.guid === childId) as Account;
    setAccountPaths(account, accounts);
  });
}
