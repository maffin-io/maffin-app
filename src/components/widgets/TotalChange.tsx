import React from 'react';
import { DateTime } from 'luxon';
import { BiTrendingDown, BiTrendingUp } from 'react-icons/bi';

import { useBalanceSheet } from '@/hooks/api';
import { useInterval } from '@/hooks/state';
import Money from '@/book/Money';
import type { Account } from '@/book/entities';

export type TotalChangeProps = {
  account: Account,
  className?: string,
};

export default function TotalChange({
  account,
  className = '',
}: TotalChangeProps): JSX.Element {
  const { data: interval } = useInterval();
  const { data: t0 } = useBalanceSheet(interval.start?.minus({ day: 1 }) as DateTime<true>);
  const { data: t1 } = useBalanceSheet(interval.end as DateTime<true>);

  const total0 = t0?.[account.guid] || new Money(0, account.commodity.mnemonic);
  const total1 = t1?.[account.guid] || new Money(0, account.commodity.mnemonic);
  const difference = total1.subtract(total0);

  return (
    <div className={`flex items-center ${className}`}>
      {
        difference.toNumber() >= 0
        && <BiTrendingUp className="mr-1 text-success" />
      }
      {
        difference.toNumber() < 0
        && <BiTrendingDown className="mr-1 text-danger" />
      }
      {difference.format()}
      {' '}
      this period
    </div>
  );
}
