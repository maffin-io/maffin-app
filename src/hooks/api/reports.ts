import React from 'react';
import {
  useQuery,
} from '@tanstack/react-query';
import { DateTime, Interval } from 'luxon';
import type { UseQueryResult } from '@tanstack/react-query';

import { aggregateChildrenTotals } from '@/helpers/accountsTotalAggregations';
import { Split } from '@/book/entities';
import { getAccountsTotals } from '@/lib/queries';
import { isAsset, isLiability } from '@/book/helpers';
import type { AccountsTotals } from '@/types/book';
import type { Account } from '@/book/entities';
import type { PriceDBMap } from '@/book/prices';
import { useAccounts } from './useAccounts';
import { usePrices } from './usePrices';
import fetcher from './fetcher';

/**
 * Calculates the balance sheet for all asset/liability/equity accounts
 * at a given point in time.
 */
export function useBalanceSheet(
  selectedDate: DateTime = DateTime.now(),
  select?: (data: AccountsTotals) => AccountsTotals,
): UseQueryResult<AccountsTotals> {
  const { data: accounts, dataUpdatedAt: accountsUpdatedAt } = useAccounts();
  const { data: prices, dataUpdatedAt: pricesUpdatedAt } = usePrices({});

  const aggregate = React.useCallback(
    ((data: AccountsTotals) => aggregateChildrenTotals(
      ['type_asset', 'type_liability'],
      accounts as Account[],
      prices as PriceDBMap,
      selectedDate,
      data,
    )),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [accountsUpdatedAt, pricesUpdatedAt, selectedDate.toISODate()],
  );

  const queryKey = [
    ...Split.CACHE_KEY,
    {
      aggregation: 'balance-sheet',
      date: selectedDate.toISODate(),
      accountsUpdatedAt,
    },
  ];
  const result = useQuery({
    queryKey,
    queryFn: fetcher(
      async () => getAccountsTotals(
        (accounts as Account[]).filter(a => a.type !== 'INCOME' && a.type !== 'EXPENSE'),
        Interval.fromDateTimes(
          DateTime.fromISO('1970'),
          selectedDate,
        ),
      ),
      queryKey,
    ),
    enabled: !!accounts,
    select: select || aggregate,
    networkMode: 'always',
  });

  return result;
}

/**
 * Calculates the income statement for all income/expense accounts
 * in a given interval of time.
 */
export function useIEStatement(
  interval: Interval,
  select?: (data: AccountsTotals) => AccountsTotals,
): UseQueryResult<AccountsTotals> {
  const { data: accounts, dataUpdatedAt: accountsUpdatedAt } = useAccounts();
  const { data: prices, dataUpdatedAt: pricesUpdatedAt } = usePrices({});

  const aggregate = React.useCallback(
    ((data: AccountsTotals) => aggregateChildrenTotals(
      ['type_income', 'type_expense'],
      accounts as Account[],
      prices as PriceDBMap,
      interval.end as DateTime<true>,
      data,
    )),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [accountsUpdatedAt, pricesUpdatedAt, interval.toISODate()],
  );

  const queryKey = [
    ...Split.CACHE_KEY,
    {
      aggregation: 'income-statement',
      interval: interval.toISODate(),
      accountsUpdatedAt,
    },
  ];
  const result = useQuery({
    queryKey,
    queryFn: fetcher(
      async () => getAccountsTotals(
        (accounts as Account[]).filter(
          a => !isAsset(a) && !isLiability(a) && a.type !== 'EQUITY',
        ),
        interval,
      ),
      queryKey,
    ),
    enabled: !!accounts,
    select: select || aggregate,
    networkMode: 'always',
  });

  return result;
}

export function useAccountsTotals(
  interval: Interval,
  select?: (data: AccountsTotals) => AccountsTotals,
): UseQueryResult<AccountsTotals> {
  const { data: balanceSheet, dataUpdatedAt: bsUpdatedAt } = useBalanceSheet(
    interval.end as DateTime<true>,
    select,
  );
  const { data: incomeStatement, dataUpdatedAt: iesUpdatedAt } = useIEStatement(interval, select);

  const queryKey = [
    ...Split.CACHE_KEY,
    {
      aggregation: 'total',
      interval: interval.toISODate(),
      bsUpdatedAt,
      iesUpdatedAt,
    },
  ];
  const result = useQuery({
    queryKey,
    queryFn: () => ({ ...balanceSheet, ...incomeStatement }),
    enabled: !!balanceSheet && !!incomeStatement,
    networkMode: 'always',
  });

  return result;
}
