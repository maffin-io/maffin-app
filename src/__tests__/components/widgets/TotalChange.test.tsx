import React from 'react';
import { Interval } from 'luxon';
import { render, screen } from '@testing-library/react';
import type { DefinedUseQueryResult, UseQueryResult } from '@tanstack/react-query';

import * as apiHook from '@/hooks/api';
import * as stateHooks from '@/hooks/state';
import Money from '@/book/Money';
import type { AccountsTotals } from '@/types/book';
import type { Account, Commodity } from '@/book/entities';
import TotalChange from '@/components/widgets/TotalChange';

jest.mock('@/hooks/api', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/api'),
}));

jest.mock('@/hooks/state', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/state'),
}));

describe('TotalChange', () => {
  let account: Account;

  beforeEach(() => {
    jest.spyOn(apiHook, 'useBalanceSheet').mockReturnValue({ data: undefined } as UseQueryResult<AccountsTotals>);
    account = {
      guid: 'guid',
      commodity: {
        mnemonic: 'EUR',
      } as Commodity,
    } as Account;
    jest.spyOn(stateHooks, 'useInterval').mockReturnValue({ data: TEST_INTERVAL } as DefinedUseQueryResult<Interval>);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders with no data', () => {
    const { container } = render(<TotalChange account={account} className="mt-1" />);

    expect(apiHook.useBalanceSheet).toBeCalledTimes(2);
    expect(
      apiHook.useBalanceSheet,
    ).toHaveBeenNthCalledWith(1, TEST_INTERVAL.start?.minus({ day: 1 }));
    expect(apiHook.useBalanceSheet).toHaveBeenNthCalledWith(2, TEST_INTERVAL.end);
    expect(container).toMatchSnapshot();
  });

  it('renders as expected with increase', () => {
    jest.spyOn(apiHook, 'useBalanceSheet')
      .mockReturnValueOnce({
        data: { guid: new Money(100, 'EUR') } as AccountsTotals,
      } as UseQueryResult<AccountsTotals>)
      .mockReturnValueOnce({
        data: { guid: new Money(200, 'EUR') } as AccountsTotals,
      } as UseQueryResult<AccountsTotals>);

    const { container } = render(<TotalChange account={account} />);

    screen.getByText('€100.00', { exact: false });
    expect(container).toMatchSnapshot();
  });

  it('renders as expected with decrease', () => {
    jest.spyOn(apiHook, 'useBalanceSheet')
      .mockReturnValueOnce({
        data: { guid: new Money(200, 'EUR') } as AccountsTotals,
      } as UseQueryResult<AccountsTotals>)
      .mockReturnValueOnce({
        data: { guid: new Money(100, 'EUR') } as AccountsTotals,
      } as UseQueryResult<AccountsTotals>);

    const { container } = render(<TotalChange account={account} />);

    screen.getByText('€100.00', { exact: false });
    expect(container).toMatchSnapshot();
  });
});
