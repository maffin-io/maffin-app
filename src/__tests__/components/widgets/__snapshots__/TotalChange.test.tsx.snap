// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TotalChange renders as expected with decrease 1`] = `
<div>
  <div
    class="flex items-center "
  >
    <svg
      class="mr-1 amount-negative"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m14 9.586-4 4-6.293-6.293-1.414 1.414L10 16.414l4-4 4.293 4.293L16 19h6v-6l-2.293 2.293z"
      />
    </svg>
    -€100.00
     
    this period
  </div>
</div>
`;

exports[`TotalChange renders as expected with increase 1`] = `
<div>
  <div
    class="flex items-center "
  >
    <svg
      class="mr-1 amount-positive"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m10 10.414 4 4 5.707-5.707L22 11V5h-6l2.293 2.293L14 11.586l-4-4-7.707 7.707 1.414 1.414z"
      />
    </svg>
    €100.00
     
    this period
  </div>
</div>
`;

exports[`TotalChange renders with no data 1`] = `
<div>
  <div
    class="flex items-center mt-1"
  >
    <svg
      class="mr-1 amount-positive"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m10 10.414 4 4 5.707-5.707L22 11V5h-6l2.293 2.293L14 11.586l-4-4-7.707 7.707 1.414 1.414z"
      />
    </svg>
    €0.00
     
    this period
  </div>
</div>
`;
